lib/hapi-proxy.js

```diff
@@ -0,0 +1,99 @@
+'use strict';
+
+const Hapi = require('hapi');
+const util = require('./util');
+
+// HapiProxy is an attempt to let developers use the standard Hapi API, but
+// get Stok's added functionality. In order to accomplish this, we proxy all
+// the necessary methods. Since Hapi doesn't expose all the internal classes
+// (e.g., `Connection` is not exposed) we have to proxy the methods
+// that expose instances (e.g., `server.connection()`) so that we can get a
+// reference to the instance and proxy the methods we actually care about.
+class HapiProxy {
+  constructor() {
+    this._server = new Hapi.Server();
+    this._isStarted = false;
+
+    this._proxyServer();
+  }
+
+  getServer() {
+    return this._server;
+  }
+
+  isStarted() {
+    return this._isStarted;
+  }
+
+  _proxyServer() {
+    this._proxyServerStart();
+
+    this._proxyServerConnection();
+    this._proxyConnection(this._server);
+  }
+
+  _proxyServerStart() {
+    util.proxy(this._server, 'start', (original) => {
+      return () => {
+
+        // `server.start()` always expects a callack. If it doesn't receive one,
+        // it will generate a callback that resolves a promise and then re-invoke
+        // the method with the generated callback. This would result in an infinite
+        // loop since it would end up invoking our proxied version, which would
+        // then invoke the original method again without a callback. To avoid this,
+        // we need to do the promise wrapping around the callback ourselves.
```

Comment:
This was fun to track down. I spent quite a while thinking there was a bug in my proxying code. 👎